### PR TITLE
Add GRU & LSTM layer to stax

### PR DESCRIPTION
**Pull request by RobertTLange**
Addresses issue #644. Adds a GRU & LSTM layer that is batch processing compatible and compiles rapidly. I first tried to work with pull request #1196 but found hick-ups regarding batch processing and compilation time (no lax.scan).

### CODE REVIEW

**Comment by RobertTLange** [comment]
@googlebot I fixed it.

File: jax/experimental/stax.py [hunk 1]
```diff
@@ -339,6 +342,81 @@ def apply_fun(params, inputs, **kwargs):
...
+        cell = np.multiply(change_gate, input_gate) + np.multiply(cell, forget_gate)
+
+        output_gate = sigmoid(np.dot(inp, out_W) + np.dot(hidden, out_U) + out_b)
+        output = np.multiply(output_gate, np.tanh(cell))
+        hidden_cell = (hidden, cell)
+        return hidden_cell, hidden_cell
+
+    def apply_fun(params, inputs, **kwargs):
+        """ Loop over the time steps of the input sequence """
+        h = params[0]
```
Comment (david-waterworth) [inline jax/experimental/stax.py]: I think it would be useful if you had the option to either learn the initial state (which is what this code does), zero it (which is the keras default for stateless) or use some sort of state initialisation function f(inputs)? Also perhaps support keras state-full where each batch uses the state from the previous.

Comment (david-waterworth) [reply jax/experimental/stax.py]: I raise this because I'm specifically researching state initialisation for RNN's (dynamical models specifically) - there are a number of options and the correct strategy has an impact on training - no doubt because the wrong initial state can results in unstable dynamics. [1] indicates that learning the initial state per batch can result in overfitting, they use a state estimator network instead.

[1] https://dx.doi.org/10.1109/ijcnn.2017.7966138

File: jax/experimental/stax.py [hunk 2]
```diff
@@ -339,6 +342,81 @@ def apply_fun(params, inputs, **kwargs):
...
+        )
+
+        k1, k2, k3 = random.split(k1, num=3)
+        change_W, change_U, change_b = (
+            W_init(k1, (input_shape[2], out_dim)),
+            W_init(k2, (out_dim, out_dim)),
+            b_init(k3, (out_dim,)),
+        )
+
+        output_shape = (input_shape[0], out_dim)
```
Comment (david-waterworth) [inline jax/experimental/stax.py]: Should this not be output_shape = (input_shape[0], input_shape[1], out_dim)? If I understand correctly the input tensor is (batches,sequences,features) so the output tensor is (batches,sequences,out_dim)?

Comment (RobertTLange) [reply jax/experimental/stax.py]: Thanks David for taking the time! Yes, you are right - I will change that asap.

File: jax/experimental/stax.py (hunk 1 above)
Comment (RobertTLange) [reply jax/experimental/stax.py]: Yes, that makes sense. I wanted to have a first working version. Adding flexibility sounds totally reasonable. Do you have any recommendations how one can keep the hidden state to be part of the params and not automatically optimize it when applying the opt update to the parameters? I have been searching for a clean way to fix/not optimize parameters for a while.

Comment (david-waterworth) [reply jax/experimental/stax.py]: First what I would change is you've created an initial state per batch. Problem with this is how do you `apply` a trained model to a test set with a different number of batches? I think what you could do is make `hidden = b_init(rng, (out_dim,))` i.e. learn a single (best average) initial state over all batches from the train data. You'll then have to tile this in apply() by the number of batches (I think, maybe the tensor product will work without tiling?).

I've been thinking a lot about adding flexibility - I've not got a good solution yet. I feel like there should be two collections returned from `init_fun()` - `params` and `states` (and both passed to `apply_fun()`). But I'm still not 100% clear on the best design as there are three general cases in my mind. Learned State (your implementation), Explicit State (i.e. zero) and Estimated State (an additional function - perhaps another jax layer - which estimates the state, this gets complex though as for some (causal) models you want to estimate state based on the previous not future inputs/outputs). 

Also looking at Keras, they have two RNN modes - stateless and statefull. The former uses constant (zero) initialisation per batch, the second holds the last state in memory (but doesn't learn it - and there's a reset method to zero it). To do this in stax I think you'd have to make states a parameter of `apply_fun` and also return the last state per batch. Then the caller can either pass in or reset as required. The problem is stax layers are single input / single output - you really want to return (outputs,state) and still be able to feed outputs into a Dense layer but have it pass through state.

So in short the "best" solution will probably require a change to the stax method signatures i.e.

```
def Layer():
  def init_fun(rng, input_shape):
    ...
    return output_shape, params, states

  def apply_fun(params, inputs, states, **kwargs):
    ...
    return outputs, states
```

Comment (david-waterworth) [reply jax/experimental/stax.py]: Also I've figured a potential (hack?) way to either have learned or explicit state initialisation that's clean(ish). If in your `apply_fun` you ignore the state from params and use a value passed in explicitly via args or kwargs the grad wrt the params will be 0 i.e.


```
    def apply_fun(params, inputs, **kwargs):

        states = kwargs.pop("states", None)
        if states is None:
            _,_,state = params
            # need to repeat initial state for each batch
            states = np.repeat(state[np.newaxis,:], inputs.shape[0], axis=0)
        #else:
            # use explicit state initialisation if passed - this will force dL/dparams[-1] == 0 is this stable?
```

File: jax/experimental/stax.py (hunk 2 above)
Comment (david-waterworth) [reply jax/experimental/stax.py]: I noticed the other stax layers do this

output_shape = input_shape[:-1] + (out_dim,)

At first I assumed it just allowed you to type less - but I've realised it's actually allows you to pass an input shape of either (sequences,features) or (batches,sequences,features) - i.e. it gives you a bit more flexibility to train on one or multiple batches which is a nice feature

File: jax/experimental/stax.py (hunk 1 above)
Comment (david-waterworth) [reply jax/experimental/stax.py]: An additional though, based on some experimentation I've been doing this week. I think there's value in implementing the RNN layers as "cells" i.e. lift the handling of the scan to an outer function. That way the RNNCell `apply_fun` simply operates on a (batches_dim,embeddings_dim) array for a single timestep. The signature would be

```
def LSTMCell(...)
  # single RNN cell (states,intputs)->(states,outputs)
  def init_fun(rng, input_shape):  # input_shape is (batches,embeddings)
    ...
    return output_shape, params

  def apply_fun(params, states, inputs, **kwargs):    # note additional input / output arg
    ...
    return states, outputs

  return init_fun, apply_fun
```

Now you can wrap this in a LSTM() layer and do the looping etc. I think this allows the LSTMCell() to be reused more easily - i.e. I want to handle variable sequence lengths, often LSTMCells are stacked etc.

ie. 

```
LSTM(...):
  # LSTM layer contains a single cell
  cell_init, cell_apply = LSTMCell(...)

  def init_fun(rng, input_shape):  # input_shape is (batches,sequences,embeddings)
    # cell state is a learnt parameter
     initial_cell_state = bias_init(...)
     output_shape=...
     return output_shape,(initial_cell_state,cell_init())

  def apply_fun(params, inputs, **kwargs):
     # need to swap batch and sequence axis to scan along sequences 
     initial_state, parms = params
     inputs = np.swapaxes(inputs, 0, 1)
     scan_fun = partial(cell_apply, params)
     _,outputs = jax.lax.scan(scan_fun, states, inputs)
     outputs = np.swapaxes(outputs, 0, 1)
     return outputs
```

What are your thoughts?

Comment (RobertTLange) [reply jax/experimental/stax.py]: Sorry David for the late response. Yes, making the layer more modular sounds like a good idea. Feel free to add whatever you find adequate. I am currently a little time limited but will return once I find time again!

**Comment by John-Boik** [comment]
Interested in the GRU implementation. Will the merge occur soon?

**Comment by sokrypton** [comment]
I was reviewing the proposed LSTM merge. It seems the output gate is not being used?
image

Unless... I'm looking at the wrong version of the code.
https://github.com/google/jax/blob/c463fe09d26497ede02e04d629ab87e3120cc529/jax/experimental/stax.py

**Comment by shoyer** [comment]
As the [README for JAX experimental](https://github.com/google/jax/tree/master/jax/experimental) states: "we generally will not merge PRs adding new features. Instead, please send your lovely pull requests and design ideas to more fully-featured libraries like Haiku, Flax, or Trax."

Both Haiku and Flax include GRU & LSTM modules -- I recommend giving them a try.
